rfc/0078-zone-aware-replica-reads.md

```diff
@@ -0,0 +1,346 @@
+        {
+            // ...
+            "hostname": "172.17.0.2",
+            "serverGroup": "group_1"
+        },
+        {
+            // ...
+            "hostname": "172.17.0.3",
+            "serverGroup": "group_2"
+        },
+        {
+            // ...
+            "hostname": "172.17.0.4",
+            "serverGroup": "group_2"
+        },
+        {
+            // ...
+            "hostname": "172.17.0.5",
+            "serverGroup": "group_1"
+        }
+    ]
+    // ...
+}
+```
+
+Having this information, the SDK will be able to filter list of the node
+indexes in `vBucketMap` to get local server group members.
+
+For example, lets say the configured SDK to use `"group_1"` as its local server
+group. Then if some key is mapped to vBucket `0`, the SDK can filter vector if
+indexes `[0, 1]` to only `[0, -1]` (using `-1` here for illustration), because
+server with index `1` belongs to `"group_2"`. In the result, the SDK only need
+to send `GET` request to retrieve active copy of the document.
+
+## Selecting Server Group for Connection
+
+To allow user to select preferred server groups, new setter/property for
+`ClusterOptions` must be introduced:
+
+```
+class ClusterOptions {
+    // ...
+    ClusterOptions preferredServerGroup(String serverGroupName);
+    // ...
+}
+```
+
+Where `serverGroupName` matches name of the group as it seen in Admin UI of the
+server and in cluster configuration JSON.
+
+## Selecting Read Preference for Operations
+
+New enumeration should be defined for expressing different strategies for the
+read replica APIs:
+
+```
+enum ReadPreference {
+    NO_PREFERENCE,
+    SELECTED_SERVER_GROUP,
+    SELECTED_SERVER_GROUP_OR_ALL_AVAILABLE,
+};
+```
+
+Each of the operation should get new option, with default value of
+`ReadPreference::NO_PREFERENCE`.
+
+```
+class GetAllReplicasOptions {
+    // ...
+    ReadPreference readPreference { ReadPreference::NO_PREFERENCE };
+    // ...
+};
+```
+
+```
+class GetAnyReplicaOptions {
+    // ...
+    ReadPreference readPreference { ReadPreference::NO_PREFERENCE };
+    // ...
+};
+```
+
+```
+class LookupInAllReplicasOptions {
+    // ...
+    ReadPreference readPreference { ReadPreference::NO_PREFERENCE };
+    // ...
+};
+```
+
+```
+class LookupInAnyReplicaOptions {
+    // ...
+    ReadPreference readPreference { ReadPreference::NO_PREFERENCE };
+    // ...
+};
+```
+
+In all failure cases, when the SDK cannot handle operation, it must return `102
+DocumentIrretrievable` with human-readable explanation whenever it is possible
+that explains the details.
+
+## Transactions
+
+Inside transaction closure, the replica read API takes simplified form, and
+always read from local group, which implies that the SDK will always throw `102
+DocumentIrretrievable` error for servers that do not expose `serverGroup`
+property in cluster configuration.
+
+```
+class TransactionAttemptContext {
+    // ...
+    public TransactionGetResult getReplica(Collection collection,
```

Comment:
Ah so maybe this has got lost in translation but I think this should be getReplicaFromPreferred[Server]Group() or similar.

(Or more accurately: I wish I had the time and capability to design a true getReplica() that handles all single-replica-read concerns.  But I cannot figure out what that should look like.  So until either inspiration strikes or I get more cycles to spend on it, the thought is to have a single-purpose method instead that is focussed on only this use-case.  That doesn't "salt the earth" from doing a true getReplica() at a later date.)